Add ConsoleCLIDownload for oadp-cli

Joeavaikath opened the pull request:
## Why the changes were made

Tackles migtools/oadp-cli#76

Two main changes:

- Add config for deployment and service pointing to image containing oadp-cli binaries hosted on a Go server
- Add a way to start the route and ConsoleCLIDownload resource on operator startup
  - This is done here because the hostname for the route is dynamic: we need to grab that once the route is up, and then create the ConsoleCLIDownload

Flow:

1. Operator deployed, new deployment and service created
2. oadp-cli controller sees deployment, creates route, backs off for hostname to be assigned
3. oadp-cli controller reconciles again, sees hostname and creates a ConsoleCLIDownload resource with this hostname as the download link


## How to test the changes made

- Run make deploy-olm
- Click the "?" symbol at the top right of the OpenShift console, and then click "Command Line Tools"
<img width="1512" height="343" alt="Screenshot 2025-10-20 at 10 01 19 AM" src="https://github.com/user-attachments/assets/c70be624-5ac9-4f17-9a74-6809ff550cfd" />

Joeavaikath committed:
Add ConsoleCLIDownload for oadp-cli

Signed-off-by: Joseph <jvaikath@redhat.com>

weshayutin commented on bundle/manifests/oadp-operator.oadp-cli.yaml:
```diff
@@ -0,0 +1,10 @@
+  displayName: oadp - OADP operator Command Line Interface (CLI)
+  links:
+    - href: https://github.com/migtools/oadp-cli/releases/
+      text: Download oadp
```
do we want s/Download oadp/Download OADP CLI/  ?

weshayutin replied:
FWIW... OADP should always be in caps

Joeavaikath replied:
So from the other cli tools on that page, I gathered that lowercase is for the actual command (oc, odo)
So lowercase is used to emphasize that it's the CLI tool which gives you access to that command on the CLI
Basically, matches the other CLI tools on that page

weshayutin commented on bundle/manifests/oadp-operator.oadp-cli.yaml:
```diff
@@ -0,0 +1,10 @@
+  description: oadp-cli is the command line tool for working with the OADP operator
+  displayName: oadp - OADP operator Command Line Interface (CLI)
+  links:
+    - href: https://github.com/migtools/oadp-cli/releases/
```
for now this is fine, but the cli needs to be built by Red Hat.. so we'll need a container in openshift-adp hosting the built artifcats.

Joeavaikath replied:
Overhauled, points to quay.io/konveyor org now
Downstream will replace the image value [here](https://github.com/openshift/oadp-operator/pull/1997/files#diff-ce0d597a9e5f20a8fe61c3ba5e6185a6b90cfef2c00bb5b5ef3e97b15b79f480R91-R92)

weshayutin commented:
This is just a step to making the oadp-cli available to customers.  Additional requiements
1. build the client for mac, windows, linux in the https://github.com/migtools/oadp-cli repo
2. push container to quay
3.  from the oadp-operator, host a container that serves the built artifacts e.g. https://downloads-openshift-console.apps.wdharm420.migration.redhat.com/ 
4. test. :)
5.  Notify @rayfordj of the new container for OADP-1.6 and point at the dockerfile

Issue tracker: https://github.com/migtools/oadp-cli/issues/71

kaovilai commented:
note that the yaml here will need to also be in config/ dir and run `make bundle`. can walk though that later if needed.

Joeavaikath committed:
Setup oadp-cli binary server on operator startup - First pass, route and ConsoleCLIDownload will be moved into a controller

Signed-off-by: Joseph <jvaikath@redhat.com>

weshayutin commented on bundle/manifests/oadp-operator.clusterserviceversion.yaml:
```diff
@@ -1190,6 +1202,40 @@ spec:
+                app: oadp-cli
+            spec:
+              containers:
+              - image: quay.io/rh_ee_jvaikath/oadp-cli-binaries
```
push the container for konveyor org

weshayutin replied:
See: https://github.com/openshift/oadp-operator/blob/oadp-dev/docs/developer/PROW_CI.md
```
Finally, add image to core-services/image-mirroring/konveyor/mapping_konveyor_latest
```

kaovilai replied:
IMO do not need prow ci to complicate things..

kaovilai replied:
We just need github actions to be authenticated and push there on commit push.

kaovilai replied:
Prow image building/mirroring should be reserved for repos where automation are not easily done transparently in GHA.

Joeavaikath replied:
Created a github workflow: on new release, directly push oadp-cli-binaries image to https://quay.io/repository/konveyor/oadp-cli-binaries?namespace=konveyor

PR: https://github.com/migtools/oadp-cli/pull/75

Joeavaikath replied:
Set to konveyor org, decided to not use prow since we're not limited by openshift org

kaovilai commented on bundle/manifests/oadp-operator.clusterserviceversion.yaml:
```diff
@@ -1190,6 +1202,40 @@ spec:
+                app: oadp-cli
+            spec:
+              containers:
+              - image: quay.io/rh_ee_jvaikath/oadp-cli-binaries
```
I think it would be neat if this container is self contained and can also create the CLIDownload CR. but if it end up being easier to do via oadp-operator manager container then that's fine too.

Joeavaikath replied:
We need some mechanism to get the hostname from the route that we create to go in to ConsoleCLIDownload
Either manager or controller

If it's packaged in the container, I think we're stuck, no way to get that dynamic hostname

kaovilai replied:
Container can have oc or kubectl which should be able to dynamically get hostname

Joeavaikath committed:
Use controller to startup route and ConsoleCLIDownload

Better than piggybacking off leader election, if leader changes it will repeatcontroller can better ensure route changes and deployment changes don't affect the end goal of having ConsoleCLIDownload with the right link

Signed-off-by: Joseph <jvaikath@redhat.com>

kaovilai commented:
ai-explains-test-failure: The unit test build succeeded, but the final **lint** step failed:

* `cmd/main.go:43:1` – the file isn’t properly formatted according to **gci** rules.
* `internal/controller/oadp_cli_controller.go:39:2` – the variable `log` shadows the imported `log` package, flagged by **revive**.

Because `golangci‑lint` reported these two errors, the job exited with status 2 and the overall step failed.

Joeavaikath committed:
Fix download link text

Signed-off-by: Joseph <jvaikath@redhat.com>

Joeavaikath committed:
Linting

Signed-off-by: Joseph <jvaikath@redhat.com>

Joeavaikath commented:
/retest

Joeavaikath committed:
Use konveyor org

Signed-off-by: Joseph <jvaikath@redhat.com>

Joeavaikath commented:
/retest-required

weshayutin reviewed (dismissed):
The image needs to be built on both AMD64 and ARM64 for dev testing purposes.

```
Node-Selectors:              <none>
Tolerations:                 node.kubernetes.io/memory-pressure:NoSchedule op=Exists
                             node.kubernetes.io/not-ready:NoExecute op=Exists for 300s
                             node.kubernetes.io/unreachable:NoExecute op=Exists for 300s
Events:
  Type     Reason          Age                    From               Message
  ----     ------          ----                   ----               -------
  Normal   Scheduled       5m45s                  default-scheduler  Successfully assigned openshift-adp/openshift-adp-oadp-cli-server-7689b54d59-dkzcv to ip-10-0-116-201.us-west-2.compute.internal
  Normal   AddedInterface  5m45s                  multus             Add eth0 [10.131.1.162/23] from ovn-kubernetes
  Normal   Pulled          5m44s                  kubelet            Successfully pulled image "quay.io/konveyor/oadp-cli-binaries" in 996ms (996ms including waiting). Image size: 353031244 bytes.
  Normal   Pulled          5m42s                  kubelet            Successfully pulled image "quay.io/konveyor/oadp-cli-binaries" in 883ms (883ms including waiting). Image size: 353031244 bytes.
  Normal   Pulled          5m28s                  kubelet            Successfully pulled image "quay.io/konveyor/oadp-cli-binaries" in 1.296s (1.296s including waiting). Image size: 353031244 bytes.
  Normal   Pulled          5m2s                   kubelet            Successfully pulled image "quay.io/konveyor/oadp-cli-binaries" in 1.302s (1.302s including waiting). Image size: 353031244 bytes.
  Normal   Pulled          4m8s                   kubelet            Successfully pulled image "quay.io/konveyor/oadp-cli-binaries" in 1.09s (1.09s including waiting). Image size: 353031244 bytes.
  Normal   Created         2m43s (x6 over 5m44s)  kubelet            Created container: oadp-cli-server
  Normal   Started         2m43s (x6 over 5m44s)  kubelet            Started container oadp-cli-server
  Normal   Pulled          2m43s                  kubelet            Successfully pulled image "quay.io/konveyor/oadp-cli-binaries" in 1.157s (1.157s including waiting). Image size: 353031244 bytes.
  Warning  BackOff         16s (x26 over 5m41s)   kubelet            Back-off restarting failed container oadp-cli-server in pod openshift-adp-oadp-cli-server-7689b54d59-dkzcv_openshift-adp(9835f1c8-309f-4bd1-bc9a-f3e0aabbc2fc)
  Normal   Pulling         2s (x7 over 5m45s)     kubelet            Pulling image "quay.io/konveyor/oadp-cli-binaries"
  Normal   Pulled          1s                     kubelet            Successfully pulled image "quay.io/konveyor/oadp-cli-binaries" in 882ms (882ms including waiting). Image size: 353031244 bytes.
whayutin@fedora:~/OPENSHIFT/git/OADP/oadp-operator$ oc get all -n openshift-adp
Warning: apps.openshift.io/v1 DeploymentConfig is deprecated in v4.14+, unavailable in v4.10000+
NAME                                                                  READY   STATUS             RESTARTS      AGE
pod/f31243f3480a6431d31a9ab844198a2fac80d19d92455907166296678bh26sq   0/1     Completed          0             6m13s
pod/openshift-adp-controller-manager-64d4c876f8-svg5g                 1/1     Running            0             6m2s
pod/openshift-adp-oadp-cli-server-7689b54d59-dkzcv                    0/1     CrashLoopBackOff   6 (17s ago)   6m2s
pod/ttl-sh-oadp-operator-bundle-8046d1af-1h                           1/1     Running            0             6m17s

NAME                                                       TYPE        CLUSTER-IP      EXTERNAL-IP   PORT(S)    AGE
service/openshift-adp-cli-server                           ClusterIP   172.30.125.71   <none>        80/TCP     6m6s
service/openshift-adp-controller-manager-metrics-service   ClusterIP   172.30.11.45    <none>        8443/TCP   6m6s

NAME                                               READY   UP-TO-DATE   AVAILABLE   AGE
deployment.apps/openshift-adp-controller-manager   1/1     1            1           6m3s
deployment.apps/openshift-adp-oadp-cli-server      0/1     1            0           6m3s

NAME                                                          DESIRED   CURRENT   READY   AGE
replicaset.apps/openshift-adp-controller-manager-64d4c876f8   1         1         1       6m3s
replicaset.apps/openshift-adp-oadp-cli-server-7689b54d59      1         1         0       6m3s

NAME                                                                        STATUS     COMPLETIONS   DURATION   AGE
job.batch/f31243f3480a6431d31a9ab844198a2fac80d19d92455907166296678b5f237   Complete   1/1           6s         6m14s

NAME                                             HOST/PORT                                                                 PATH   SERVICES                   PORT   TERMINATION     WILDCARD
route.route.openshift.io/oadp-cli-server-route   oadp-cli-server-route-openshift-adp.apps.wdharm420.migration.redhat.com          openshift-adp-cli-server   http   edge/Redirect   None
whayutin@fedora:~/OPENSHIFT/git/OADP/oadp-operator$ oc logs -f deployment.apps/openshift-adp-oadp-cli-server
exec container process `/usr/local/bin/download-server`: Exec format error

```

weshayutin commented:
<img width="552" height="1286" alt="Screenshot From 2025-10-29 13-42-19" src="https://github.com/user-attachments/assets/42cc3675-e01b-48ba-8eb7-304a91f67d90" />
listed twice

weshayutin commented:
Installation should be tightened up a little:
```
Installation:

tar -xzf kubectl-oadp_*.tar.gz
chmod +x kubectl-oadp
sudo mv kubectl-oadp /usr/local/bin/
```
TO
```
tar -xzf kubectl-oadp_*.tar.gz
either 
  - Fix the unpack to be the binary name kubectl-oadp
  - OR
  - mv kubectl-oadp_v*_linux_$arch kubectl-oadp
 sudo mv kubectl-oadp /usr/local/bin/
```

Joeavaikath committed:
Start oadp-cli on server startup

Signed-off-by: Joseph <jvaikath@redhat.com>

Joeavaikath commented:
ConsoleCLIDownload is a cluster scoped resource
We can't set an Owner reference so that it gets deleted when the operator is uninstalled
So this makes it idempotent, we won't have multiples just updates the existing ConsoleCLIDownload on another install
To clean it up we would need a manual delete

kaovilai commented:
> To clean it up we would need a manual delete

As long as each time this creates, if one already exists, it gets updated to the newest version without failing I think it's safe to have this be leftover.

kaovilai commented on bundle/manifests/oadp-operator.clusterserviceversion.yaml:
```diff
@@ -1135,6 +1147,8 @@ spec:
                 - name: RELATED_IMAGE_NON_ADMIN_CONTROLLER
                   value: quay.io/konveyor/oadp-non-admin:latest
+                - name: RELATED_IMAGE_CONSOLE_CLI_DOWNLOAD
+                  value: quay.io/konveyor/console-cli-download:latest
```
bundle/config is out of date. run `make bundle` to update this value from manager.yaml

Joeavaikath committed:
Make bundle

Signed-off-by: Joseph <jvaikath@redhat.com>

Joeavaikath commented:
On a fresh operator install, the deployment is reset and so we get the latest image for oadp-cli-binaries regardless of ConsoleCLIDownload resource being there, cause it just points to it
Now, do we want to update it continuously somehow is another question: i.e. actively look for a new version
edit: maybe its best not to disturb versions like that, not reproducable
we could settle on a particular version and update it on our end

kaovilai commented:
since its synced to operator version, just run once at the operator startup.

kaovilai commented:
but make sure if there exists existing, with a diff val but same name, that creation/update succeed on getting the right version served.

Joeavaikath commented:
/retest-failed

Joeavaikath commented:
/retest-required

Joeavaikath commented:
/retest-required

Joeavaikath commented:
/retest-required

Joeavaikath commented:
/retest

Joeavaikath commented:
/retest-required

kaovilai commented:
/retest

ai-retester: The `e2e-test-aws-e2e` step failed because the test suite failed. Specifically, the "Mongo application CSI" test timed out while waiting for a pod to succeed. The todolist pod was stuck in the PodInitializing state.

kaovilai commented:
/retest

ai-retester: The `e2e-test-cli-aws` test failed because the `Mongo application CSI via CLI` test timed out while waiting for the `todolist` pod to become ready. The container "todolist" in pod "todolist-6d7bb9554c-ntgnx" is waiting to start: PodInitializing

kaovilai commented:
/retest

ai-retester: The `e2e-test-aws` step failed because the `Mongo application CSI` test timed out while waiting for the `todolist` pod to become ready. The container in the pod is waiting to start/initialize.

kaovilai commented:
/retest

ai-retester: The "Mongo application CSI" test timed out because the `todolist` pod failed to initialize, preventing successful backup and restore.

kaovilai commented:
/retest

ai-retester: The end-to-end test `e2e-test-aws` failed because the `Mongo application CSI` test timed out after 540 seconds while waiting for a Pod to succeed, and the test suite thus failed.

kaovilai commented:
/retest

ai-retester: The Mongo backup/restore test never got out of the “waiting to start” state – the `todolist` pod hung in **PodInitializing** because the scheduler rejected it for violating the cluster’s PodSecurity policy (the container was marked as privileged, used unsupported securityContext flags, and the pod had no required fields). The test timed out (~540 s) waiting for the pod to succeed, so it failed.

kaovilai commented:
/retest

ai-retester: The `e2e-test-cli-aws-e2e` step failed because the `Mongo application CSI via CLI` test timed out while waiting for the `todolist` pod to become ready during the OADP end-to-end testing. The "container "todolist" in pod ... is waiting to start: PodInitializing" message confirms, the pod never reached active status for testing.

kaovilai commented:
/retest

weshayutin commented:
@kaovilai @Joeavaikath you have to rebase :)

kaovilai commented:
As far as prow goes, no you don't. :) assuming no conflicts

weshayutin commented:
/retest

kaovilai commented:
> As far as prow goes, no you don't. :) assuming no conflicts

receipts if interested in digging more:

https://github.com/kubernetes-sigs/prow/blob/8cd4af136cc0eeb9ee7ab9fc11cd99f7adba663e/pkg/pod-utils/clone/clone.go#L359

> [Before a PR is merged, Tide ensures that all jobs configured as required in the presubmits part of the config.yaml file are **passing against the latest base branch commit, rerunning the jobs if necessary**. No job is required to be configured in which case it’s enough if a PR meets all GitHub search criteria.](https://docs.prow.k8s.io/docs/components/core/tide/config/#:~:text=Example-,Configuring%20Presubmit%20Jobs,-Before%20a%20PR)

emphasis mine.

TL;DR: each test run contains changes from base branch merged into it. Subsequent retests or new test triggers also bring latest base branch changes into it.

kaovilai commented:
/retest

ai-retester: The `e2e-test-cli-aws` test failed because the "MySQL application two Vol CSI via CLI" test case failed, with a pod initialization error, preventing the application from being accessible. The underlying issue appears to be that the "ebs.csi.aws.com" volume snapshots were not ready when the restore was attempted, causing provisioning to fail.

kaovilai commented:
/retest

ai-retester: The `e2e-test-cli-aws` test failed because the `MySQL application two Vol CSI via CLI` test case failed. The Todolist service could not be reached after restore, likely due to issues with the pod initializing despite the restore completing otherwise. The underlying cause could be due to CSI snapshot restore failures in the persistent data location because, while the restore was marked as being completed at 2025-11-07 17:15:22 +0000 UTC, the service wasn't accessible at 2025-11-07 17:16:02.128.

Joeavaikath commented:
/retest-required

Joeavaikath committed:
Do not initialize controller if watchNamespace is empty

Signed-off-by: Joseph <jvaikath@redhat.com>

Joeavaikath committed:
Add bounded retries, exponential backoff and context awareness

Signed-off-by: Joseph <jvaikath@redhat.com>

kaovilai commented:
/retest

Joeavaikath commented:
All tests passing, wow

Joeavaikath commented on internal/controller/cli_download_controller.go:
```diff
@@ -0,0 +1,390 @@
+			Port: &routev1.RoutePort{
+				TargetPort: intstr.FromString("http"),
+			},
+			TLS: &routev1.TLSConfig{
```
Test after removing this
Traffic from route to svc is http

Joeavaikath replied:
The TLSConfig on lines 380-383 does two things:

  1. TLSTerminationEdge - Enables HTTPS for external traffic
  2. InsecureEdgeTerminationPolicyRedirect - Forces HTTP users to upgrade to HTTPS

"http" is just the port name, it matches the service defined "http" port

kaovilai commented:
/retest

ai-retester: The end-to-end (e2e) tests for OADP, particularly the "MySQL application two Vol CSI via CLI" test, failed due to a connectivity issue. The test was unable to connect to the service via a route and then failed even when trying to connect via a proxy pod, resulting in a "ContainerFailed" error and a failing test suite, leading to the overall job failure.

comment for openshift/oadp-operator/pull/1997

weshayutin commented:
/retest

weshayutin commented:
<img width="1046" height="545" alt="Screenshot From 2026-01-07 10-48-43" src="https://github.com/user-attachments/assets/880011c6-feb4-43f4-aad0-437cd0ddcdf7" />

Joeavaikath commented:
/retest

kaovilai commented on internal/controller/cli_download_controller.go:
```diff
@@ -0,0 +1,390 @@
+		return fmt.Errorf("failed to get CLI server route: %w", err)
+	}
+
+	// Check if route has hostname - retry with backoff if not assigned
+	hostname := route.Spec.Host
+	if hostname == "" {
+		c.Log.Info("Route hostname not yet assigned, retrying with backoff")
+		maxRetries := 5
+		backoff := 2 * time.Second
+
+		for attempt := 1; attempt <= maxRetries; attempt++ {
+			select {
+			case <-ctx.Done():
+				return ctx.Err()
+			case <-time.After(backoff):
+				// Re-fetch route to check for hostname
+				if err := c.Client.Get(ctx, client.ObjectKey{
+					Name:      cliServerRouteName,
+					Namespace: c.Namespace,
+				}, route); err != nil {
+					c.Log.Error(err, "Failed to get route on retry", "attempt", attempt)
+					continue
+				}
+
+				if route.Spec.Host != "" {
+					hostname = route.Spec.Host
+					c.Log.Info("Route hostname assigned", "hostname", hostname, "attempt", attempt)
+					break
+				}
+
+				c.Log.Info("Route hostname still not assigned, will retry", "attempt", attempt, "maxRetries", maxRetries)
+				backoff *= 2 // Exponential backoff
+			}
+		}
+
+		if hostname == "" {
+			c.Log.Info("Route hostname not assigned after max retries, ConsoleCLIDownload will not be created. It will be created on next reconciliation when hostname becomes available.")
+			return nil
+		}
+	}
```
I feel like this could be using a [Watch](https://book-v3.book.kubebuilder.io/reference/watching-resources). Only after hostname is populated (autogerated or user manual input) do we create ConsoleCLIDownload object. Could ignore.

weshayutin commented:
/LGTM

kaovilai commented:
/override ci/prow/4.21-e2e-test-aws
